deducing-this.md

```diff
@@ -366,15 +391,20 @@ It also works as a more obvious way to teach how `std::bind` and `std::function`
 
 ## ABI implications for `std::function` and related
 
-If references and pointers do not have the same representation for member functions, this effectively says "for the purposes of the `this`-designated first parameter, they do."
+The representation of references is not defined. This leads to the following
+constraint: if references and pointers do not have the same representation for member functions, this effectively says "for the purposes of the `this`-designated first parameter, they do."
 
 This matters because code written in the "`this` is a pointer" syntax with the `this->` notation needs to be assembly-identical to code written with the `self.` notation; the two are just different ways to implement a function with the same signature.
 
-## Interplays with capturing `[this]` and `[*this]`
+Note that since `this` is a pointer, one cannot overoad `operator->` for it, and
+therefore `this->` and `self.` are always equivalent expressions (at least until
+we get `operator.`)
```

Comment:
Well, question. What does this return:

    struct B {
        int i;
        
        template <typename Self>
        auto f1(Self&& this) {
            return i; // or this->i
        }

        template <typename Self>
        auto f2(Self&& this self) {
            return self.i;
        }
    };

    struct D : B {
        double i;
    };

    D d;
    auto i1 = d.f1();
    auto i2 = d.f2();

I'd suggest that `f1` should return `B::i` while `f2` has to return `D::i`.

Reply:
I strongly disagree on the principles of following our rules:
- 1) what would an inline friend function do (WWIFFD)
- 2) *this and self always refer to the same object

In order to demonstrate what I mean (and exactly how bad that turns out to be): https://godbolt.org/g/Bri2gs

I had to change the . notation to friend call notation, but it's the same thing.